pkg/server/init.go

```diff
@@ -390,6 +390,24 @@ func (s *initServer) startJoinLoop(ctx context.Context, stopper *stop.Stopper) e
 		return ErrJoinRPCUnsupported
 	}
 
+	// Busy-loop through all the resolvers at least once. Keep this code block
+	// roughly in sync with the one below.
+	for _, res := range s.config.resolvers {
+		addr := res.Addr()
+		err := s.attemptJoin(ctx, addr)
+		if err == nil {
+			return nil
+		}
+
+		if errors.Is(err, ErrJoinRPCUnsupported) || errors.Is(err, ErrIncompatibleBinaryVersion) {
```

Comment:
you should handle `ctx.Done()` and `ctx.Err()` here too

Reply:
That was an intentional omission. This API is a bit crufty as written (I've just added a TODO for what needs to change), but basically as written it needs to return nil errors on context cancellation. Which is what we're doing below. Here I just wanted to avoid doing a non-nil error check and return a nil error anyway.

Reply:
Yea there were some pretty crappy hacks in there, so I'll wait for https://github.com/cockroachdb/cockroach/pull/53934 to land and have this sit on top of that instead.

Reply:
ok I'll let you rebase and then we can wrap this up

Reply:
Done, PTAL.